### PR TITLE
Return more specific errors about invalid descriptors

### PR DESCRIPTION
**Pull request by achow101**
Because it's nigh impossible to figure out what is wrong with a descriptor otherwise.

### CODE REVIEW

**Comment by sipa** [comment]
Descriptors are hard to parse, they should be hard to write also.

Eh, I mean, big concept ACK.

**Comment by jb55** [comment]
Concept ACK

**Comment by Sjors** [comment]
Concept ACK

**Review by promag — COMMENTED** [review]
Concept ACK, nice to have a detailed error.

**Comment by hugohn** [comment]
Concept ACK.

File: src/script/descriptor.cpp [hunk 1]
```diff
@@ -696,14 +696,17 @@ NODISCARD bool ParseKeyPath(const std::vector<Span<const char>>& split, KeyPath&
             hardened = true;
         }
         uint32_t p;
-        if (!ParseUInt32(std::string(elem.begin(), elem.end()), &p) || p > 0x7FFFFFFFUL) return false;
+        if (!ParseUInt32(std::string(elem.begin(), elem.end()), &p) || p > 0x7FFFFFFFUL) {
+            error = strprintf("Key path value %u is out of range", p);
```
Comment (Sjors) [inline src/script/descriptor.cpp]: Nit: "not valid" instead of "out of range"? Or split the check?

File: test/functional/wallet_importmulti.py [hunk 2]
```diff
@@ -552,7 +552,7 @@ def run_test(self):
                                "keys": [key.privkey]},
                               success=False,
                               error_code=-5,
-                              error_message="Descriptor is invalid")
+                              error_message="Descriptor is invalid, Missing checksum")
```
Comment (Sjors) [inline test/functional/wallet_importmulti.py]: nit: make descriptor errors start with lower case?

**Review by Sjors — APPROVED** [review]
ACK dae2590

This introduces a fun little trick if you're too lazy to call `getdescriptorinfo` to obtain the checkum: you can instead set the checksum to `00000000`.
```
$ bitcoin-cli importmulti '[{"desc": "sh(wpkh(03fff97bd5755eeea420453a14355235d382f6472f8568a18b2f057a1460297556))#00000000", "timestamp": "now"}]'
[
  {
    "success": false,
    "error": {
      "code": -5,
      "message": "Descriptor is invalid, Provided checksum '00000000' does not match computed checksum 'qkrrc7je'"
    }
  }
]

```

File: test/functional/wallet_importmulti.py (hunk 2 above)
Comment (achow101) [reply test/functional/wallet_importmulti.py]: Nah.

File: src/script/descriptor.cpp (hunk 1 above)
Comment (achow101) [reply src/script/descriptor.cpp]: I've split the check

File: src/test/descriptor_tests.cpp [hunk 3]
```diff
@@ -13,13 +13,15 @@
 
 namespace {
 
-void CheckUnparsable(const std::string& prv, const std::string& pub)
+void CheckUnparsable(const std::string& prv, const std::string& pub, std::string expected_error)
```
Comment (practicalswift) [inline src/test/descriptor_tests.cpp]: Nit: `expected_error` could be const ref?

File: src/script/descriptor.cpp [hunk 4]
```diff
@@ -696,33 +696,60 @@ NODISCARD bool ParseKeyPath(const std::vector<Span<const char>>& split, KeyPath&
             hardened = true;
         }
         uint32_t p;
-        if (!ParseUInt32(std::string(elem.begin(), elem.end()), &p) || p > 0x7FFFFFFFUL) return false;
+        if (!ParseUInt32(std::string(elem.begin(), elem.end()), &p)) {
+            error = strprintf("Key path value '%s' is not a valid uint32", std::string(elem.begin(), elem.end()).c_str());
```
Comment (promag) [inline src/script/descriptor.cpp]: Missing test.

File: src/script/descriptor.cpp [hunk 5]
```diff
@@ -696,33 +696,60 @@ NODISCARD bool ParseKeyPath(const std::vector<Span<const char>>& split, KeyPath&
...
-            if (pubkey.IsFullyValid() && (permit_uncompressed || pubkey.IsCompressed())) return MakeUnique<ConstPubkeyProvider>(pubkey);
+            if (pubkey.IsFullyValid()) {
+                if (permit_uncompressed || pubkey.IsCompressed()) {
+                    return MakeUnique<ConstPubkeyProvider>(pubkey);
+                } else {
+                    error = "Uncompressed keys are not allowed";
+                    return nullptr;
+                }
+            }
+            error = strprintf("Pubkey '%s' is invalid", str);
```
Comment (promag) [inline src/script/descriptor.cpp]: Missing test.

File: src/script/descriptor.cpp [hunk 6]
```diff
@@ -741,95 +768,154 @@ std::unique_ptr<PubkeyProvider> ParsePubkeyInner(const Span<const char>& sp, boo
...
-    if (origin_split.size() > 2) return nullptr;
-    if (origin_split.size() == 1) return ParsePubkeyInner(origin_split[0], permit_uncompressed, out);
-    if (origin_split[0].size() < 1 || origin_split[0][0] != '[') return nullptr;
+    if (origin_split.size() > 2) {
+        error = "Multiple ']' characters found for a single pubkey";
+        return nullptr;
+    }
+    if (origin_split.size() == 1) return ParsePubkeyInner(origin_split[0], permit_uncompressed, out, error);
+    if (origin_split[0].size() < 1 || origin_split[0][0] != '[') {
+        error = strprintf("Key origin expected but not found, got '%s' instead", std::string(origin_split[0].begin(), origin_split[0].end()));
```
Comment (promag) [inline src/script/descriptor.cpp]: Missing test.

File: src/script/descriptor.cpp [hunk 7]
```diff
@@ -741,95 +768,154 @@ std::unique_ptr<PubkeyProvider> ParsePubkeyInner(const Span<const char>& sp, boo
...
+        error = "Cannot have combo in non-top level";
+        return nullptr;
     }
     if (Func("multi", expr)) {
         auto threshold = Expr(expr);
         uint32_t thres;
         std::vector<std::unique_ptr<PubkeyProvider>> providers;
-        if (!ParseUInt32(std::string(threshold.begin(), threshold.end()), &thres)) return nullptr;
+        if (!ParseUInt32(std::string(threshold.begin(), threshold.end()), &thres)) {
+            error = strprintf("multi threshold %u out of range", thres);
```
Comment (promag) [inline src/script/descriptor.cpp]: Missing test.

File: src/script/descriptor.cpp [hunk 8]
```diff
@@ -741,95 +768,154 @@ std::unique_ptr<PubkeyProvider> ParsePubkeyInner(const Span<const char>& sp, boo
...
+            return nullptr;
+        }
         size_t script_size = 0;
         while (expr.size()) {
-            if (!Const(",", expr)) return nullptr;
+            if (!Const(",", expr)) {
+                error = strprintf("multi: expected ',', got '%c'", expr[0]);
```
Comment (promag) [inline src/script/descriptor.cpp]: Missing test.

File: src/script/descriptor.cpp [hunk 9]
```diff
@@ -741,95 +768,154 @@ std::unique_ptr<PubkeyProvider> ParsePubkeyInner(const Span<const char>& sp, boo
...
             auto arg = Expr(expr);
-            auto pk = ParsePubkey(arg, ctx != ParseScriptContext::P2WSH, out);
+            auto pk = ParsePubkey(arg, ctx != ParseScriptContext::P2WSH, out, error);
             if (!pk) return nullptr;
             script_size += pk->GetSize() + 1;
             providers.emplace_back(std::move(pk));
         }
-        if (providers.size() < 1 || providers.size() > 16 || thres < 1 || thres > providers.size()) return nullptr;
+        if (providers.size() < 1 || providers.size() > 16) {
+            error = strprintf("Cannot have %u keys in multisig; must have between 1 and 16 keys, inclusive", providers.size());
```
Comment (promag) [inline src/script/descriptor.cpp]: Missing test.

File: src/script/descriptor.cpp [hunk 10]
```diff
@@ -741,95 +768,154 @@ std::unique_ptr<PubkeyProvider> ParsePubkeyInner(const Span<const char>& sp, boo
...
+            return nullptr;
+        } else if (thres < 1) {
+            error = strprintf("multisig threshold cannot be %d, must be at least 1", thres);
```
Comment (promag) [inline src/script/descriptor.cpp]: Missing test.

File: src/script/descriptor.cpp [hunk 11]
```diff
@@ -741,95 +768,154 @@ std::unique_ptr<PubkeyProvider> ParsePubkeyInner(const Span<const char>& sp, boo
...
+            return nullptr;
+        } else if (thres > providers.size()) {
+            error = strprintf("Multisig threshold cannot be larger than the number of keys; thoreshold is %d but only %u keys specified", thres, providers.size());
```
Comment (promag) [inline src/script/descriptor.cpp]: Missing test.

File: src/script/descriptor.cpp [hunk 12]
```diff
@@ -741,95 +768,154 @@ std::unique_ptr<PubkeyProvider> ParsePubkeyInner(const Span<const char>& sp, boo
...
+            return nullptr;
+        }
         if (ctx == ParseScriptContext::TOP) {
-            if (providers.size() > 3) return nullptr; // Not more than 3 pubkeys for raw multisig
+            if (providers.size() > 3) {
+                error = strprintf("Cannot %u pubkeys in bare multisig; only at most 3 pubkeys", providers.size());
```
Comment (promag) [inline src/script/descriptor.cpp]: Missing test.

File: src/script/descriptor.cpp [hunk 13]
```diff
@@ -741,95 +768,154 @@ std::unique_ptr<PubkeyProvider> ParsePubkeyInner(const Span<const char>& sp, boo
...
         return MakeUnique<WSHDescriptor>(std::move(desc));
+    } else if (ctx == ParseScriptContext::P2WSH && Func("wsh", expr)) {
+        error = "Cannot have wsh within wsh";
+        return nullptr;
     }
     if (ctx == ParseScriptContext::TOP && Func("addr", expr)) {
         CTxDestination dest = DecodeDestination(std::string(expr.begin(), expr.end()));
-        if (!IsValidDestination(dest)) return nullptr;
+        if (!IsValidDestination(dest)) {
+            error = "Address is not valid";
```
Comment (promag) [inline src/script/descriptor.cpp]: Missing test.

File: src/script/descriptor.cpp [hunk 14]
```diff
@@ -741,95 +768,154 @@ std::unique_ptr<PubkeyProvider> ParsePubkeyInner(const Span<const char>& sp, boo
...
+            return nullptr;
+        }
         return MakeUnique<AddressDescriptor>(std::move(dest));
     }
     if (ctx == ParseScriptContext::TOP && Func("raw", expr)) {
         std::string str(expr.begin(), expr.end());
-        if (!IsHex(str)) return nullptr;
+        if (!IsHex(str)) {
+            error = "Raw script is not hex";
```
Comment (promag) [inline src/script/descriptor.cpp]: Missing test.

File: src/script/descriptor.cpp [hunk 15]
```diff
@@ -910,23 +996,38 @@ std::unique_ptr<DescriptorImpl> InferScript(const CScript& script, ParseScriptCo
...
+std::unique_ptr<Descriptor> Parse(const std::string& descriptor, FlatSigningProvider& out, std::string& error, bool require_checksum)
 {
     Span<const char> sp(descriptor.data(), descriptor.size());
 
     // Checksum checks
     auto check_split = Split(sp, '#');
-    if (check_split.size() > 2) return nullptr; // Multiple '#' symbols
-    if (check_split.size() == 1 && require_checksum) return nullptr; // Missing checksum
+    if (check_split.size() > 2) {
+        error = "Multiple '#' symbols";
```
Comment (promag) [inline src/script/descriptor.cpp]: Missing test.

File: src/script/descriptor.cpp [hunk 16]
```diff
@@ -910,23 +996,38 @@ std::unique_ptr<DescriptorImpl> InferScript(const CScript& script, ParseScriptCo
...
-        if (check_split[1].size() != 8) return nullptr; // Unexpected length for checksum
+        if (check_split[1].size() != 8) {
+            error = strprintf("Expected 8 character checksum, not %u characters", check_split[1].size());
+            return nullptr;
+        }
         auto checksum = DescriptorChecksum(check_split[0]);
-        if (checksum.empty()) return nullptr; // Invalid characters in payload
-        if (!std::equal(checksum.begin(), checksum.end(), check_split[1].begin())) return nullptr; // Checksum mismatch
+        if (checksum.empty()) {
+            error = "Invalid characters in payload";
```
Comment (promag) [inline src/script/descriptor.cpp]: Missing test.

File: test/functional/wallet_importmulti.py (hunk 2 above)
Comment (promag) [reply test/functional/wallet_importmulti.py]: Agree with @Sjors, is there a reason to not do so? I was going to suggest to drop "Descriptor is invalid, " prefix.

**Review by promag — COMMENTED** [review]
I think I've bookmarked all errors without a test - not saying you should add them here, I'm happy to improve in a follow up.

File: src/script/descriptor.cpp (hunk 8 above)
Comment (achow101) [reply src/script/descriptor.cpp]: I don't think this check is actually possible to hit due to how `Expr` parses things earlier.

**Comment by achow101** [comment]
I've added more tests and also cleaned up a few typos in error messages.

File: src/test/descriptor_tests.cpp (hunk 3 above)
Comment (achow101) [reply src/test/descriptor_tests.cpp]: Done

File: test/functional/wallet_importmulti.py (hunk 2 above)
Comment (achow101) [reply test/functional/wallet_importmulti.py]: I've just removed the prefixes.

**Comment by Sjors** [comment]
reACK a1b47955838f1151fc1904d0ff3bc7751b9fa7ca

**Comment by jb55** [comment]
Code Review ACK a1b47955838f1151fc1904d0ff3bc7751b9fa7ca

File: src/script/descriptor.cpp [hunk 17]
```diff
@@ -710,17 +713,35 @@ std::unique_ptr<PubkeyProvider> ParsePubkeyInner(const Span<const char>& sp, boo
...
-            CPubKey pubkey = key.GetPubKey();
-            out.keys.emplace(pubkey.GetID(), key);
-            return MakeUnique<ConstPubkeyProvider>(pubkey);
+        if (key.IsValid()) {
+            if (permit_uncompressed || key.IsCompressed()) {
+                CPubKey pubkey = key.GetPubKey();
+                out.keys.emplace(pubkey.GetID(), key);
+                return MakeUnique<ConstPubkeyProvider>(pubkey);
+            } else {
+                error = "Uncompressed keys are not allowed";
```
Comment (instagibbs) [inline src/script/descriptor.cpp]: where would this case be hit previously?

File: src/script/descriptor.cpp [hunk 18]
```diff
@@ -798,19 +819,22 @@ std::unique_ptr<DescriptorImpl> ParseScript(Span<const char>& sp, ParseScriptCon
         auto pubkey = ParsePubkey(expr, true, out, error);
         if (!pubkey) return nullptr;
         return MakeUnique<ComboDescriptor>(std::move(pubkey));
+    } else if (ctx != ParseScriptContext::TOP && Func("combo", expr)) {
```
Comment (instagibbs) [inline src/script/descriptor.cpp]: where would this case be hit previously?

File: src/script/descriptor.cpp [hunk 19]
```diff
@@ -866,6 +908,13 @@ std::unique_ptr<DescriptorImpl> ParseScript(Span<const char>& sp, ParseScriptCon
         auto bytes = ParseHex(str);
         return MakeUnique<RawDescriptor>(CScript(bytes.begin(), bytes.end()));
     }
+    if (ctx == ParseScriptContext::P2SH) {
```
Comment (instagibbs) [inline src/script/descriptor.cpp]: nit: I'd prefer it also checks for `!Func` explicitly, since there are a number of combinations, and a generic error is preferable to a specific error that's an implementation mistake.

**Review by instagibbs — APPROVED** [review]
tACK

the errors I encountered during initial usage of descriptors are all caught here, and enough information to quickly figure out what I did wrong.

f.e.

```
getdescriptorinfo "multi(1, 03a34b99f22c790c4e36b2b3c2c35a36db06226e41c692fc82b8b56ac1c540c5bd,04a34b99f22c790c4e36b2b3c2c35a36db06226e41c692fc82b8b56ac1c540c5bd5b8dec5235a0fa8722476c7709c02559e3aa73aa03918ba2d492eea75abea235)"
error code: -5
error message:
key ' 03a34b99f22c790c4e36b2b3c2c35a36db06226e41c692fc82b8b56ac1c540c5bd' is not valid
```
^ space after comma

File: src/script/descriptor.cpp (hunk 17 above)
Comment (achow101) [reply src/script/descriptor.cpp]: It would fail below at the extkey decoding stuff.

File: src/script/descriptor.cpp (hunk 18 above)
Comment (achow101) [reply src/script/descriptor.cpp]: It would just get to the end of the function and hit `return nullptr`.

**Comment by Sjors** [comment]
If you need to make any changes, I suggest rebasing due to #14934 to be on the safe side. It still compiles and tests pass for me when I do that rebase locally though.

File: src/script/descriptor.cpp (hunk 19 above)
Comment (achow101) [reply src/script/descriptor.cpp]: I don't think it can check that there is no function.

**Comment by meshcollider** [comment]
Sorry about the rebase again due to #15986 

Once this is rebased I'll review and merge

**Comment by achow101** [comment]
Rebased

File: src/script/descriptor.cpp (hunk 1 above)
Comment (Sjors) [reply src/script/descriptor.cpp]: Now it's no longer split.

File: test/functional/wallet_importmulti.py (hunk 2 above)
Comment (Sjors) [reply test/functional/wallet_importmulti.py]: They're back

**Review by Sjors — CHANGES_REQUESTED** [review]
Only the first commit changed in this rebase, but you lost a few of your recent fixes in the process.

**Comment by achow101** [comment]
> Only the first commit changed in this rebase, but you lost a few of your recent fixes in the process.

No??? Are you sure you're looking at the right commits? The latest diff includes everything that was commented on earlier.

File: test/functional/wallet_importmulti.py (hunk 2 above)
Comment (achow101) [reply test/functional/wallet_importmulti.py]: > They're back

No??

File: src/script/descriptor.cpp (hunk 1 above)
Comment (achow101) [reply src/script/descriptor.cpp]: > Now it's no longer split.

No??

**Comment by Sjors** [comment]
Ah wait, I see these changes happen in 625534d7b1417da926f1ced600855ea818d6e01e.

ACK 787c9ec

File: src/script/descriptor.cpp [hunk 20]
```diff
@@ -760,7 +781,7 @@ std::unique_ptr<PubkeyProvider> ParsePubkey(const Span<const char>& sp, bool per
     }
     if (origin_split.size() == 1) return ParsePubkeyInner(origin_split[0], permit_uncompressed, out, error);
     if (origin_split[0].size() < 1 || origin_split[0][0] != '[') {
-        error = strprintf("Key origin expected but not found, got '%s' instead", std::string(origin_split[0].begin(), origin_split[0].end()));
+        error = strprintf("Key origin start '[ character expected but not found, got '%c' instead", origin_split[0][0]);
```
Comment (meshcollider) [inline src/script/descriptor.cpp]: mismatched `'` (also in the corresponding test case)

**Review by meshcollider — COMMENTED** [review]
Code review ACK 787c9ec0c383cefb83a95433311a27f9c220dca0

This is a nice change, one tiny nit but it is pretty insignificant so I'll merge this tomorrow if you don't have time to fix it now
